### ISSUE TITLE
runtime: network blips during concurrent GC

### ISSUE THREAD

**Issue by petermattis** [golang/go#20457]
### What version of Go are you using (`go version`)?

`go version go1.8.1 darwin/amd64`

### What operating system and processor architecture are you using (`go env`)?

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/pmattis/Development/go"
GORACE=""
GOROOT="/Users/pmattis/Development/go-1.8"
GOTOOLDIR="/Users/pmattis/Development/go-1.8/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/qc/fpqpgdqd167c70dtc6840xxh0000gn/T/go-build955668609=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```

### What did you do?

While investigating tail latencies in [CockroachDB](https://github.com/cockroachdb/cockroach), I noticed network blips during concurrent GC. The trace below is from a single-node cockroach cluster with a single client sending it requests as fast as possible. Note that the load is modest. My test machine is ~75% idle while gathering these traces.

![screen shot 2017-05-22 at 11 25 20 am](https://cloud.githubusercontent.com/assets/6201335/26316211/81235200-3ee1-11e7-884e-402df9114204.png)

[trace.8.out.zip](https://github.com/golang/go/files/1019460/trace.8.out.zip)

Every time GC runs, there are no networking events for ~10ms which is quite close to the length of the GC run itself. This behavior occurs very quickly after starting the cluster and load generator. The heap size is modest (O(100 MB)).

Here is a zoom in on the first GC run from the above trace:

![screen shot 2017-05-22 at 12 54 18 pm](https://cloud.githubusercontent.com/assets/6201335/26319731/228fdc88-3eee-11e7-903c-fdcfbbfa93bc.png)

In addition to the blip while GC is running, there is a smaller blip after each GC run while sweeping. The image below shows G611 being called on to assist in sweep work from 62ms-69ms in the trace.

![screen shot 2017-05-22 at 1 16 31 pm](https://cloud.githubusercontent.com/assets/6201335/26320521/0e681362-3ef1-11e7-8a5c-7e406a07a05f.png)

Note that I'm running both the `cockroach` process and the load generator ([`kv`](https://github.com/cockroachdb/loadgen/tree/master/kv)) on the same machine. Are the GC runs in the `cockroach` process starving the `kv` process of CPU? Running both processes with `GOMAXPROCS=4` (I'm using an 8 core machine), shows much better behavior.

![screen shot 2017-05-22 at 11 25 58 am](https://cloud.githubusercontent.com/assets/6201335/26316214/839b910a-3ee1-11e7-9a30-24a54adac053.png)

[trace.4.out.zip](https://github.com/golang/go/files/1019462/trace.4.out.zip)

**Comment by dsnet** [comment]
\cc @aclements

**Comment by ianlancetaylor** [comment]
I note that the `sysmon` goroutine polls the network every 10ms if nothing else has polled.  Normally `findrunnable` will poll the network whenever it doesn't have another goroutine to run.  Perhaps during the GC there is always another goroutine to run (to do GC work) and therefore we fall back to the `sysmon` polling.

**Comment by petermattis** [comment]
I forgot to add that I instrumented the Go runtime and can see lots of calls to `netpoll` via `gcDrain` -> `pollWork`. I also tried lowering the max sleep time in `sysmon` to 1ms, but that had no effect. Perhaps the 10ms is some OS level time slice on Darwin. I'll try and verify this same behavior on Linux.

**Comment by petermattis** [comment]
Linux shows similar behavior:

<img width="1259" alt="screen shot 2017-05-22 at 8 56 14 pm" src="https://cloud.githubusercontent.com/assets/6201335/26334216/793e986a-3f31-11e7-8175-b26bfa9193da.png">

[trace.linux.out.zip](https://github.com/golang/go/files/1020610/trace.linux.out.zip)

I'm not sure what to make of the 3 idle procs (5, 6 and 7).

**Comment by josharian** [comment]
Probably unrelated, but #20307 also observed that GC had a detrimental effect on goroutine scheduling.

**Comment by josharian** [comment]
@petermattis if you can reproduce easily, would you mind trying out tip (and maybe 1.7.5) to see whether anything has changed?

**Comment by petermattis** [comment]
@josharian Tip as of `go version devel +5e79787 Mon May 22 13:44:12 2017 +0000 darwin/amd64`:

<img width="1256" alt="screen shot 2017-05-22 at 9 43 24 pm" src="https://cloud.githubusercontent.com/assets/6201335/26335119/d13b9d3c-3f37-11e7-9525-bf0513ea5005.png">

[trace.tip.out.zip](https://github.com/golang/go/files/1020671/trace.tip.out.zip)

The networking blip is still there during GC, but it is no longer precisely 10ms. The network blips during sweep seem to be much reduced:

<img width="1332" alt="screen shot 2017-05-22 at 9 45 38 pm" src="https://cloud.githubusercontent.com/assets/6201335/26335169/10c13976-3f38-11e7-87e1-81887cf17ee4.png">

**Comment by petermattis** [comment]
I'm reasonably convinced that what I'm seeing is CPU starvation of my load generator. I tweaked the output of `GODEBUG=gctrace=1` to compute the amount of CPU used by GC during the concurrent mark phase:

```
gc 203 @28.154s 0/98%: 0.007+9.5+0.039 ms clock, 0.060+0.21/18/56+0.31 ms cpu, 82->82->42 MB, 84 MB goal, 8 P
```
The `0/98%` indicates that GC has used 0% of the CPU over the lifetime of the process, but used 98% of the CPU during the 9.5ms of concurrent marking. The use of the idle CPU for GC is reasonable on the one hand, but makes the Go process a poor neighbor.

I tried disabling idle GC workers which significantly reduced the CPU spikes during GC, but made latencies much worse:

![screen shot 2017-05-24 at 9 49 48 am](https://cloud.githubusercontent.com/assets/6201335/26406494/6a8fd456-4066-11e7-9970-54be401d1e51.png)

[trace.noidlegc.out.zip](https://github.com/golang/go/files/1025729/trace.noidlegc.out.zip)

In the above trace, `G377` is running on `Proc 5` when GC starts. `G377` then blocks on network IO and the 2 dedicated GC workers get started. But there are only 2 Procs running and there does not seem to be a mechanism to kick off another one (or perhaps I disabled that mechanism in my hack to disable idle GC workers). Adding a `wakep` call after a dedicated GC worker is started fixes this issue. This gets latencies to a better place:

![screen shot 2017-05-24 at 9 58 33 am](https://cloud.githubusercontent.com/assets/6201335/26406897/a72bc8a6-4067-11e7-920c-345cb92493b8.png)

[trace.noidlegc2.out.zip](https://github.com/golang/go/files/1025769/trace.noidlegc2.out.zip)

CPU usage during marking is much lower than stock go1.8.1:

```
gc 201 @30.960s 1/29%: 0.025+22+0.080 ms clock, 0.20+11/41/0+0.64 ms cpu, 87->88->45 MB, 89 MB goal, 8 P
```

The downside to removing the idle GC workers is that the GC periods are about twice as long.

Seems like there is a high-level question to be answered about how much CPU GC should use when the process is not CPU limited.

**Comment by bradfitz** [comment]
Dup of #17969 ("aggressive GC completion is disruptive to co-tenants") then?

**Comment by petermattis** [comment]
Yep, that seems like the same issue I'm seeing.

**Comment by aclements** [comment]
> But there are only 2 Procs running and there does not seem to be a mechanism to kick off another one (or perhaps I disabled that mechanism in my hack to disable idle GC workers).

You're right that at the moment it's possible for not all idle Ps to help with GC. This is #14179 (which we fixed once, but the fix was broken, so then we unfixed it).

> Seems like there is a high-level question to be answered about how much CPU GC should use when the process is not CPU limited.

Yes. :) But I don't know what the answer is. Ideally we'd get the OS scheduler involved, since it's the only thing that knows what CPU resources are *really* idle. But (right now) we need global coordination between the GC workers, so we can't just run them at OS idle priority.

Thanks for all the traces and debugging! Closing as a dup of #17969.
